docs/content/docs/try-flink/write_flink_program_with_sql.md

```diff
@@ -0,0 +1,262 @@
+distributed with this work for additional information
+regarding copyright ownership.  The ASF licenses this file
+to you under the Apache License, Version 2.0 (the
+"License"); you may not use this file except in compliance
+with the License.  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing,
+software distributed under the License is distributed on an
+"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+KIND, either express or implied.  See the License for the
+specific language governing permissions and limitations
+under the License.
+-->
+
+# Write your first Flink program with SQL
+
+## Introduction
+
+Flink features [multiple APIs]({{< ref "docs/concepts/overview" >}}) with different levels of abstraction that can be used to develop your streaming application. SQL is the highest level of abstraction and is supported by Flink as a relational API for batch and stream processing. This means that you can write the same queries on both unbounded real-time streams and bounded recorded streams and produce the same results. 
+
+SQL on Flink is based on [Apache Calcite](https://calcite.apache.org/) (which is based on standard SQL) and is commonly used to ease the process of implementing data analytics, data pipelining, and ETL applications.  It is a great entryway to writing your first Flink application and requires no Java or Python. 
+
+This tutorial will guide you through writing your first Flink program leveraging SQL alone. Through this exercise you will learn and understand the ease and speed with which you can analyze streaming data in Flink! 
+
+
+## Goals
+
+This tutorial will teach you how to:
+
+- use the Flink SQL client to submit queries 
+- consume a data source with Flink SQL
+- run a continuous query on a stream of data
+- use Flink SQL to write out results to persistent storage 
+
+
+## Prerequisites 
+
+You only need to have basic knowledge of SQL to follow along.
+
+
+## Step 1: Start the Flink SQL client 
+
+The [SQL Client]({{< ref "docs/dev/table/sqlClient" >}}) is bundled in the regular Flink distribution and can be run out-of-the-box. It requires only a running Flink cluster where table programs can be executed (since Flink SQL is a thin abstraction over the Table API). 
+
+There are many ways to set up Flink but you will run it locally for the purpose of this tutorial. [Download Flink]({{< ref "docs/try-flink/local_installation#downloading-flink" >}}) and [start a local cluster]({{< ref "docs/try-flink/local_installation#starting-and-stopping-a-local-cluster" >}}) with one worker (the TaskManager).  
+
+The scripts for the SQL client are located in the `/bin` directory of Flink. You can start the client by executing:
+
+```sh
+./bin/sql-client.sh
+```
+
+You should see something like this:
+
+{{< img src="/fig/try-flink/flink-sql.png" alt="Flink SQL client" >}}
+
+
+## Step 2: Set up a data source with flink-faker
```

Comment:
for a beginner tutorial, the user needs to download many things. wouldn't it be better to use Flink's datagen which is built-in and doesn't require additional steps.

Reply:
I wanted to find a balance between realistic and simplicity.  Do you think that the datagen connector is as flexible as flink-faker and can create realistic examples?

Reply:
If you really wanna do a realistic example, then I suggest to go with a full fledged connector like kafka or filesystem, instead of flink-faker, so you can also show the concept of connector and format.

Reply:
I fear if we use datagen the examples will be significantly less compelling.

Reply:
@slinkydeveloper I get where you're coming from, but using real connectors is problematic.

With Kafka we'd either have to force folks to install kafka (not gonna do that), use docker (also not very workable, unfortunately), or we somehow set up an embedded kafka (again, no). Plus we'd need to somehow generate data into kafka. Besides, we already have examples based on docker-ized kafka, and they haven't worked out very well.

Using the filesystem might be a great idea, but I believe we want this to be a streaming example, and afaik we don't yet have a streaming filesystem table source.

Reply:
> Using the filesystem might be a great idea, but I believe we want this to be a streaming example, and afaik we don't yet have a streaming filesystem table source. Glad you're working on https://issues.apache.org/jira/browse/FLINK-20286.

Exactly, and the streaming filesystem table source should be ready for the next release.

Another thing I personally don't like is that flink-faker is not an official Apache project, but instead it's a community project, hence I don't feel comfortable using it here in the official docs.

I personally prefer that either you use datagen or connector files, as IMHO flink-faker is not a good connector example.

Reply:
I don't see how to create a good streaming SQL demo/tutorial experience on 1.14 without using flink-faker.

Let's assume we'll have a filesystem table source in 1.15 -- how will we use it in tutorials like this one? I'm guessing we can somehow generate a file containing a non-trivial amount of interesting data, and then host that file somewhere. Would we put that at some convenient URI -- maybe in an S3 bucket, or something similar -- and configure a table that points directly there?

Reply:
Personally, I am in favor of keeping flink-faker for now to get the piece published and then iterating on it.  Even though it is not an official Apache project, it was created by someone from Ververica and I feel like it doesn't need that much maintenance based on what it does.  Although I could be wrong. :p

Reply:
An alternative would be to switch over to using a file now, and rework the tutorial to be batch only until 1.15, after which it could be expanded to show both batch and streaming.

But I'm still unsure about the logistics of using a file. Ideally such a file should be hosted somewhere under the control of the project -- so maybe in some github repo.

Reply:
I think even a simple file committed here is enough, something like a single csv file or 1000 lines

Reply:
We use the flink-faker extensively in the Ververica SQL recipes. I would +1 for using the flink-faker here too. You could even make an argument that for proper tutorials like this, we should have a flink-faker in the Flink repo itself (but that's another discussion)